Fix labels and annotations propagation to k8s service on update

dsimansk opened the pull request:
Fixes #15891

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Fix labels and annotations propagation to k8s service on update

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Fix labels and annotations propagation to k8s service on update
```

/cc @dprotaso @Fedosin 

At least a few tests would be useful here. I'll add those based on initial feedback.

dprotaso commented:
test failures look legit

dprotaso commented:
This behaviour should have been captured in conformance tests as per the spec - https://github.com/knative/specs/blob/main/specs/serving/knative-api-specification-1.0.md#service-1

Want to add a test here - https://github.com/knative/serving/tree/main/test/conformance/api/v1

dsimansk commented:
/test istio-latest-no-mesh

dsimansk commented:
/test istio-latest-no-mesh-tls

dsimansk commented:
> test failures look legit

Fixed that to preserve existing labels and annotations on k8s service.


> This behaviour should have been captured in conformance tests as per the spec - https://github.com/knative/specs/blob/main/specs/serving/knative-api-specification-1.0.md#service-1
> 
> Want to add a test here - https://github.com/knative/serving/tree/main/test/conformance/api/v1

The spec doesn't explicitly mention Kubernetes Service and labels propagation. Should we amend that to the docs? I'll take a look at adding conformance tests to cover it.

dprotaso commented:
@dsimansk it's already in the [spec](https://github.com/knative/specs/blob/main/specs/serving/knative-api-specification-1.0.md#service-1)

> Updates to [Service] spec, metadata.labels, and metadata.annotations of the Service MUST be copied to the appropriate Configuration or Route,

So we just need to update conformance

dprotaso commented on pkg/reconciler/route/reconcile_resources.go:
```diff
@@ -256,10 +257,25 @@ func (c *Reconciler) updatePlaceholderServices(ctx context.Context, route *v1.Ro
+					existing = from.Service.DeepCopy()
 					existing.Spec = to.Service.Spec
+				}
+				if !equality.Semantic.DeepEqual(from.Service.Annotations, kmap.Union(from.Service.Annotations, to.Service.Annotations)) {
```
This reads a bit awkward - can we simplify the if structure to just one eg:

```go
if spec changed || annotations changed || labels changed {
 // copy spec, annotations, and labels
}
```

Fedosin replied:
Agree with @dprotaso - it's better to do something like this:

```
   if !equality.Semantic.DeepEqual(from.Service.Spec, to.Service.Spec) ||
      !equality.Semantic.DeepEqual(from.Service.Annotations, kmap.Union(from.Service.Annotations, to.Service.Annotations)) ||
      !equality.Semantic.DeepEqual(from.Service.Labels, kmap.Union(from.Service.Labels, to.Service.Labels)) {
       // Don't modify the informers copy.
       existing := from.Service.DeepCopy()
       existing.Spec = to.Service.Spec
       existing.Annotations = kmap.Union(from.Service.Annotations, to.Service.Annotations)
       existing.Labels = kmap.Union(from.Service.Labels, to.Service.Labels)
       if _, err := c.kubeclient.CoreV1().Services(ns).Update(ctx, existing, metav1.UpdateOptions{}); err != nil {
           return err
       }
   }
```

Overall the code looks good and solves the issue.

dsimansk commented:
@dprotaso @Fedosin this should be ready for next round. I've simplified the change check per suggestions, thanks. And added conformance tests for annotations (extended existing) and labels.

Fedosin reviewed (approved):
/lgtm

dsimansk commented:
@dprotaso any objection to cherry-pick to 1.18 and 1.17 release branches?

dprotaso commented:
@dsimansk nope cherry pick away
